### PR TITLE
feat(draw_sw): add image clip_radius and mask before transformation

### PR DESCRIPTION
**Pull request by kisvegabor**
It required a larger restructuring in image drawing so I'm mainly relying on the CI.

### Notes
- Update the [Documentation](https://github.com/lvgl/lvgl/tree/master/docs) if needed.
- Add [Examples](https://github.com/lvgl/lvgl/tree/master/examples) if relevant.
- Add [Tests](https://github.com/lvgl/lvgl/blob/master/tests/README.md) if applicable.
- If you added new options to `lv_conf_template.h` run [lv_conf_internal_gen.py](https://github.com/lvgl/lvgl/blob/master/scripts/lv_conf_internal_gen.py) and update [Kconfig](https://github.com/lvgl/lvgl/blob/master/Kconfig).
- Run `scripts/code-format.py` (`astyle v3.4.12` needs to installed by running `cd scripts; ./install_astyle.sh`) and follow the [Code Conventions](https://docs.lvgl.io/master/CODING_STYLE.html).
- Mark the Pull request as [Draft](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/changing-the-stage-of-a-pull-request) while you are working on the first version, and mark is as _Ready_ when it's ready for review.
- When changes were requested, [re-request review](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/requesting-a-pull-request-review) to notify the maintainers.
- Help us to review this Pull Request! Anyone can [approve or request changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/approving-a-pull-request-with-required-reviews).

### CODE REVIEW

File: tests/ref_imgs/draw/c_array_image_stride1_UNCOMPRESSED_recolor.png
Comment (kisvegabor) [inline tests/ref_imgs/draw/c_array_image_stride1_UNCOMPRESSED_recolor.png:1]: @XuNeo The RGB565A8 image is broken here, however I've noticed that in its [C file](https://github.com/lvgl/lvgl/blob/master/tests/test_images/stride_align1/UNCOMPRESSED/test_RGB565A8_NONE_align1.c#L120-L122) the width and stride are inconsistent. 

If I generate the file again it seems working, but wonder what is the best way of generated all test files again.

Comment (kisvegabor) [reply tests/ref_imgs/draw/c_array_image_stride1_UNCOMPRESSED_recolor.png:1]: I converted the project mages manually, and found the issue. Now it works well.

**Review by MarkIoanAmat — APPROVED** [review]

File: src/core/lv_refr.c [hunk 1]
```diff
@@ -1113,6 +1113,10 @@ static void refr_obj(lv_layer_t * layer, lv_obj_t * obj)
             layer_area_act.y2 = layer_area_act.y1 + max_rgb_row_height - 1;
             if(layer_area_act.y2 > layer_area_full.y2) layer_area_act.y2 = layer_area_full.y2;
             bool area_need_alpha = alpha_test_area_on_obj(obj, &layer_area_act);
+
+            const void * bitmap_mask_src = lv_obj_get_style_bitmap_mask_src(obj, 0);
+            if(bitmap_mask_src) area_need_alpha  = true;
+
```
Comment (liamHowatt) [inline src/core/lv_refr.c]: Can we avoid calling `alpha_test_area_on_obj` if we know `bitmap_mask_src != NULL`?

```suggestion
            const void * bitmap_mask_src = lv_obj_get_style_bitmap_mask_src(obj, 0);
            bool area_need_alpha = bitmap_mask_src || alpha_test_area_on_obj(obj, &layer_area_act);
```

**Comment by kisvegabor** [comment]
@FASTSHIFT I've implemented bg_image radius clipping and added some tests. I've noticed that VG-Lite clips them incorrectly:
- [SW](https://github.com/lvgl/lvgl/pull/7244/files#diff-2b7b9468034d99ecbac42b38c6b129c5421909df7e17a807c10a9b705fe2fe3f)
- [VG-Lite](https://github.com/lvgl/lvgl/pull/7244/files#diff-242f305692436f289db7b3cdda6298157b9ab35835506b88b28a86b1524051abR336)
- I use[ image area](https://github.com/lvgl/lvgl/pull/7244/files#diff-242f305692436f289db7b3cdda6298157b9ab35835506b88b28a86b1524051abR336) for the clip area

**Comment by FASTSHIFT** [comment]
> @FASTSHIFT I've implemented bg_image radius clipping and added some tests. I've noticed that VG-Lite clips them incorrectly:
> 
> * [SW](https://github.com/lvgl/lvgl/pull/7244/files#diff-2b7b9468034d99ecbac42b38c6b129c5421909df7e17a807c10a9b705fe2fe3f)
> * [VG-Lite](https://github.com/lvgl/lvgl/pull/7244/files#diff-242f305692436f289db7b3cdda6298157b9ab35835506b88b28a86b1524051abR336)
> * I use[ image area](https://github.com/lvgl/lvgl/pull/7244/files#diff-242f305692436f289db7b3cdda6298157b9ab35835506b88b28a86b1524051abR336) for the clip area

OK, I'll try to fix it sometime.

**Comment by kisvegabor** [comment]
I'd still merge it and VG-Lite could be fixed later.

**Comment by kisvegabor** [comment]
Suggestion approved :slightly_smiling_face:

**Review by liamHowatt — DISMISSED** [review]

**Comment by kisvegabor** [comment]
Merge conflict resolved, @FASTSHIFT @XuNeo please take a look at this PR.

**Review by liamHowatt — APPROVED** [review]

**Review by FASTSHIFT — APPROVED** [review]

**Review by uLipe — APPROVED** [review]

**Pull request merged**